Implement outline around fonts

swehner opened the pull request:
We're using fonts with an outline in our foosball UI (white with black outline) to make sure they're readable on any color background.
It basically draws the same text moved by a few pixels with the outline color first and then on top with the foreground color - not the most efficient one as it redraws the same string several times, but enough for our purposes.
We have it in copied classes in our code, but maybe this could be a useful addition to the Font and FixedString classes.
Let me know what you think and if you want to pull in this change...

![image](https://cloud.githubusercontent.com/assets/273236/22406519/d3332880-e654-11e6-9b5f-39b0b20c69a2.png)

BTW - I'm not exactly sure why the self.image[:,:,:3] = color[:3] is necessary (it obviously breaks the outline in our case).
Wouldn't it make more sense to create the image with background (color[0], color[1], color[2], 0) if background is None - or is there some problem with PIL changing the pixel's color?

paddywwoof commented:
Hi, thanks for this. The alpha setting to the background alpha was an attempt to stop darker edges when writing light coloured characters on a light background. However it doesn't really seem to have any effect though I'm sure it did at the time it was implemented - maybe something else fixed the problem subsequent to that! I will take it out anyway.

It occurred to me (and probably to you initially) that one of the PIL filters would have done the job - CONTOUR looks especially likely - but not seem to have the functionality required. Then I thought of drawing the same FixedString twice. The first time blurred and in the outline colour, then smooth and in the foreground colour. I think this works nicely, not as dark as your outline but looks cleaner and more regular (and only two draws). What do you think? It could probably be added to the FixedString class quite easily.
![screenshot from 2017-01-29 20-36-32](https://cloud.githubusercontent.com/assets/1939398/22407721/e8790430-e663-11e6-924b-433d2d9d832c.jpg)

swehner commented:
The blurring looks decent enough - like you said it's not as dark, but it's still very readable.
Definitely a good alternative: probably faster and less code!
Did you use the BLUR filter in the end?

paddywwoof commented:
Hi, yes that's just done using two of the existing FixedString objects with two draws but I will modify the code so it can take an outline shade and, if set, generate the Texture using PIL BLUR. I will fix the alpha setting at the same time.

swehner commented:
Hi,

Ok - that makes sense, but I'd call it a shadow instead of an outline.

So for FixedString your approach of simply making two instances is actually
doable right now as it already supports the filters...
I still thinks it would made sense to add this feature to FixedString if we
add it to Font, too.

I can update the Font to use this method. The font is a bit trickier
because of the size increase of the glyphs, and now it has to render them
in two passes, first the shadow and then the proper glyph, but it should be
fine - not very different from what it is doing now.

Are you ok with the changes in generating the glyph-map?

I still have to parameterize the row_margin used - we had to use it because
PIL seemed to be slightly short on glyph dimensions and they would overlap
slightly sometimes.
I also have update the docstring and add a demo with a few different fonts
to see everything is working correctly.


On Jan 30, 2017 00:47, "paddywwoof" <notifications@github.com> wrote:

> Hi, yes that's just done using two of the existing FixedString objects
> with two draws but I will modify the code so it can take an outline shade
> and, if set, generate the Texture using PIL BLUR. I will fix the alpha
> setting at the same time.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tipam/pi3d/pull/208#issuecomment-275956432>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAQrVM2wLgoH8VvBQktCHaaTaAJ1UMlsks5rXSUjgaJpZM4Lw3AA>
> .
>

paddywwoof commented:
Yes, I agree on all those points. PIL/Pillow does seem a bit vague on glyph dimensions and positions - I had to push a mod (several years ago) in order for it to be able to draw one character strings without the base line jumping up and down.

If you are going to make these mods that would be fantastic (and probably efficient as you are already immersed in the issues) but I would be happy to do it if you want.

One thing that occurred to me: it is possible to use ImageFilter.GaussianBlur(radius=.. which would allow the size argument to be used. And yes 'shadow' is a more appropriate description.

swehner commented:
Gaussian blur sounds good!
Don't worry, I'll take care of it and update the PR with the discussed
changes

On Jan 30, 2017 11:06, "paddywwoof" <notifications@github.com> wrote:

Yes, I agree on all those points. PIL/Pillow does seem a bit vague on glyph
dimensions and positions - I had to push a mod (several years ago) in order
for it to be able to draw one character strings without the base line
jumping up and down.

If you are going to make these mods that would be fantastic (and probably
efficient as you are already immersed in the issues) but I would be happy
to do it if you want.

One thing that occurred to me: it is possible to use
ImageFilter.GaussianBlur(radius=.. which would allow the size argument to
be used. And yes 'shadow' is a more appropriate description.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/tipam/pi3d/pull/208#issuecomment-276023141>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAQrVBnHBIUJ9KbyNZReGmc-thmM_bcQks5rXbZAgaJpZM4Lw3AA>
.

swehner commented:
I've updated the PR, but honestly I'm not too happy with it...
There are quite a few things that aren't that good: The PIL bug with the dark edges you had already fixed by overwriting the colors makes composing the text and its shadow a bit more complicated (including conversions from Pil Img to np array and back).

BTW: I have reverted the changes in laying out the font image, because it breaks fonts used in TextBlock.

The shadow used in Font is not that good because it is cropped to the size of the letter - it's not *that* bad, but it is visible in a few places...

![image](https://cloud.githubusercontent.com/assets/273236/22621996/a20bec98-eb30-11e6-8b5a-87c6c531897d.png)

For example in the "w" here - it also gets more noticeable the bigger the blur radius is.
I've pushed a few samples here: https://github.com/swehner/pi3d_demos/tree/shadow_demos
Let me know what you think - it is a bit specific and has quite some issues so maybe it's not worth including it.

I might take a look at pycairo for rendering text and see if it does a better job with the transparency (and it actually supports setting stroke color).
Just for fun to see if that could be an alternative...

paddywwoof commented:
Hi, thanks for this. I don't know why the space is so tight around some of the letters - it's obviously a typographical, artistic thing for kerning. However for the purposes of Font we can afford to have a few more pixels either side - I had noticed the bottom of g and possibly y getting trimmed sometimes. I've added three on left and right and two to the height - I think it looks ok. Generally just a shadow radius of one makes the letters stand out against any background.

And I've put your mods into the develop branch for the next release. Thanks again.

swehner commented:
OK, sounds good! I'll try it out with the latest changes! Thanks

swehner commented:
I've just tested the font with the latest changes, and the difference in size is pretty noticeable.
And using it with the String class it's actually a bit awkward.

![image](https://cloud.githubusercontent.com/assets/273236/22905350/41ef1ef4-f240-11e6-8772-ee6b9d044027.png)

This is a before & after - Not sure what you think, but IMO it's a bit too much.
What do you think about reducing the extra space (or even making it a parameter?)

paddywwoof commented:
Yes, you're right! It did pass through my mind (to make it a variable, 
never a good idea to put in 'magic' numbers) When I look at the demos (I 
added the shadow effect to Shapes.py) it does look a significantly less 
tidy than the FixedString demo (which doesn't have the extra padding) So 
I've just put in an argument in Font, 'spacing' defaulting to None that 
will increase with shadow_radius. It looks better (imho). Pushed to 
develop branch.


Thanks,

Paddy


diff --git a/pi3d/util/Font.py b/pi3d/util/Font.py
index 3c065b2..229eab4 100644
--- a/pi3d/util/Font.py
+++ b/pi3d/util/Font.py
@@ -29,7 +29,7 @@ class Font(Texture):
    def __init__(self, font, color=(255,255,255,255), codepoints=None,
                 add_codepoints=None, font_size=42, image_size=512,
                 italic_adjustment=1.1, background_color=None,
-               shadow=(0,0,0,255), shadow_radius=0,
+               shadow=(0,0,0,255), shadow_radius=0, spacing=None,
                 mipmap=True, filter=None):
      """Arguments:
      *font*:
@@ -86,6 +86,11 @@ class Font(Texture):
      *shadow_radius*:
        Gaussian blur radius applied to shadow layer, default 0 (no shadow)

+    *spacing*:
+      Extra spacing between letters to allow for shadow. The default value
+      None will add spacing equal to the shadow_radius, this will be 
overridden
+      by any value supplied.
+
      *mipmap*:
        Resulting texture mipmap option, default true

@@ -105,7 +110,9 @@ class Font(Texture):
        raise Exception(msg)

      ascent, descent = imgfont.getmetrics()
-    self.height = ascent + descent + 3 # allow a few more pixels for 
bottom of g,y etc
+    if spacing is None:
+      spacing = shadow_radius
+    self.height = ascent + descent + spacing # allow extra pixels if 
shadow or for certain fonts
      self.spacing = 64

      image_size = self.spacing  * 16  # or 1024 TODO this may go wrong 
if self.height != 64
@@ -156,8 +163,8 @@ class Font(Texture):
        draw.text((curX + offset, curY), ch, font=imgfont, fill=color)
        if is_draw_shadows:
          shadow_draw.text((curX + offset, curY), ch, font=imgfont, 
fill=shadow)
-      chwidth += 6 # make a little more room (for w for instance)
-      offset -= 3
+      chwidth += spacing * 2 # make a little more room (for w for instance)
+      offset -= spacing

        x = float(curX + offset) / self.ix
        y = float(curY + self.height) / self.iy


On 13/02/17 22:02, swehner wrote:
>
> I've just tested the font with the latest changes, and the difference 
> in size is pretty noticeable.
> And using it with the String class it's actually a bit awkward.
>
> image 
> <https://cloud.githubusercontent.com/assets/273236/22905350/41ef1ef4-f240-11e6-8772-ee6b9d044027.png>
>
> This is a before & after - Not sure what you think, but IMO it's a bit 
> too much.
> What do you think about reducing the extra space (or even making it a 
> parameter?)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub 
> <https://github.com/tipam/pi3d/pull/208#issuecomment-279537201>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AB2XxjQGGK3mY7u-Vrn1gtlXTBDALQ3Uks5rcNLigaJpZM4Lw3AA>.
>

swehner commented:
Awesome - much better :) thanks!
